File: frontend/src/services/roles.ts
import { fetchApi } from '../utils/api';

export interface Permission {
  id: number;
  name: string;
  module: string | null;
  action: string | null;
  description: string | null;
}

export interface Role {
  id: number;
  name: string;
  display_name: string;
  description: string | null;
  is_active: boolean;
  permissions: Permission[];
  created_at?: string;
  updated_at?: string;
}

export interface RoleCreate {
  name: string;
  display_name: string;
  description?: string;
  is_active?: boolean;
  permissions: number[]; // Array of Permission IDs
}

export interface RoleUpdate {
  name?: string;
  display_name?: string;
  description?: string;
  is_active?: boolean;
  permissions?: number[]; // Array of Permission IDs
}

export const rolesService = {
  getAllRoles: async (): Promise<Role[]> => {
    return await fetchApi('/roles/');
  },

  getAllPermissions: async (): Promise<Permission[]> => {
    return await fetchApi('/roles/permissions');
  },

  createRole: async (data: RoleCreate): Promise<Role> => {
    return await fetchApi('/roles/', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  updateRole: async (id: number, data: RoleUpdate): Promise<Role> => {
    return await fetchApi(`/roles/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  },

  deleteRole: async (id: number): Promise<void> => {
    return await fetchApi(`/roles/${id}`, {
      method: 'DELETE',
    });
  }
};

File: backend/src/schemas/roles.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional
from datetime import datetime

class PermissionResponse(BaseModel):
    id: int
    name: str
    module: Optional[str] = None
    action: Optional[str] = None
    description: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

class RoleBase(BaseModel):
    name: str
    display_name: str
    description: Optional[str] = None
    is_active: bool = True

class RoleCreate(RoleBase):
    permissions: List[int] = []

class RoleUpdate(BaseModel):
    name: Optional[str] = None
    display_name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    permissions: Optional[List[int]] = None

class RoleResponse(RoleBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    permissions: List[PermissionResponse] = []

    model_config = ConfigDict(from_attributes=True)
